src/shims/time.rs

```diff
@@ -1,8 +1,99 @@
+use std::sync::atomic::AtomicU64;
 use std::time::{Duration, Instant, SystemTime};
 
+use rustc_data_structures::sync::Ordering;
+
 use crate::concurrency::thread::Time;
 use crate::*;
 
+const NANOSECOND_PER_BASIC_BLOCK: u64 = 10;
+
+/// A monotone clock used for `Instant` simulation.
+#[derive(Debug)]
+pub enum Clock {
+    Host {
+        /// The "time anchor" for this machine's monotone clock.
+        time_anchor: Instant,
+    },
+    Virtual {
+        /// The "current virtual time".
+        nanoseconds: AtomicU64,
+    },
+}
+
+impl Clock {
+    /// Create a new clock based on the availability of communication with the host.
+    pub fn new(communicate: bool) -> Self {
+        if communicate {
+            Self::Host { time_anchor: Instant::now() }
+        } else {
+            Self::Virtual { nanoseconds: 0.into() }
+        }
+    }
+
+    /// Get the current time relative to this clock.
+    pub fn get(&self) -> Duration {
+        match self {
+            Self::Host { time_anchor } => Instant::now().saturating_duration_since(*time_anchor),
+            Self::Virtual { nanoseconds } =>
+                Duration::from_nanos(nanoseconds.load(Ordering::Relaxed)),
+        }
+    }
+
+    /// Let the time pass for a small interval.
+    pub fn tick(&self) {
+        match self {
+            Self::Host { .. } => {
+                // Time will pass without us doing anything.
+            }
+            Self::Virtual { nanoseconds } => {
+                nanoseconds.fetch_add(NANOSECOND_PER_BASIC_BLOCK, Ordering::Relaxed);
+            }
+        }
+    }
+
+    /// Sleep for the desired duration.
+    pub fn sleep(&self, duration: Duration) {
+        match self {
+            Self::Host { .. } => std::thread::sleep(duration),
+            Self::Virtual { nanoseconds } => {
+                // Just pretend that we have slept for some time.
+                nanoseconds.fetch_add(duration.as_nanos().try_into().unwrap(), Ordering::Relaxed);
+            }
+        }
+    }
+
+    /// Compute `now + duration` relative to this clock.
+    pub fn checked_add_since_now(&self, duration: Duration) -> Option<Time> {
```

Comment:
Oh now I get it. I would suggest `get_time_relative` as name for this method, and `get_time_absolute` for the other one.